quakemigrate/io/data.py

```diff
@@ -413,6 +398,92 @@ def __init__(self, starttime, endtime, sampling_rate, stations=None,
         self.availability = None
         self.wa_waveforms = None
 
+    def check_availability(self, st, all_channels=False, allow_gaps=False,
+                           full_timespan=True):
+        """
+        Check waveform availability against data quality criteria. There are a
+        number of hard-coded checks: for whether any data is present; for
+        whether the data is a flatline (all samples have the same value); and
+        for whether the data contains overlaps. There are a selection of
+        additional optional checks which can be specified according to the
+        onset function / user preference.
+
+        Parameters
+        ----------
+        st : `obspy.Stream` object
+            Stream containing the waveform data to check against the
+            availability criteria.
+        all_channels : bool, optional
+            Whether all supplied channels (distinguished by SEED id) need to
+            meet the availability criteria to mark the data as 'available'.
+        allow_gaps : bool, optional
+            Whether to allow gaps.
+        full_timespan : bool, optional
+            Whether to ensure the data covers the entire timespan requested;
+            note that this implicitly requires that there be no gaps.
+
+        Returns
+        -------
+        available : int
+            0 if data doesn't meet the availability requirements; 1 if it does.
+        availability : dict
+            Dict of {tr_id : available} for each unique SEED ID in the input
+            stream (available is again 0 or 1).
+
+        """
+
+        availability = {}
+
+        # Check if any channels in stream
+        if bool(st):
+            # Loop through channels with unique SEED id's
+            for tr_id in sorted([tr.id for tr in st]):
```

Comment:
This trace id is just used to drop traces later, right?

Reply:
Don't quite follow... I split the stream into traces with unique ID's to enable the 'all_channels' kwarg to work (so if e.g. E is gappy for a station, but N is fine, then available can still = 1), and also in returning the availability as a dict (so that in this situation the information may be used to remove traces from E, if desired -- or fill with zeros, etc.)

Reply:
As in, this trace_id (which presumably is a full SEED ID, right?) is used to identify the traces and to create the 'availability' that is used in `onset` to remove any traces that failed the critera.

Reply:
* and only used there - but yeah, I understand what it does.

Reply:
Yep. It's also used here to get `st_id` - necessary to do by `st.select` because there might be multiple traces with the same id.